### PR TITLE
[DDO-2999] Daily rotation window

### PR DESCRIPTION
**Pull request by choover-broad**
Add new flags to yale, `-window-start=HH:MM` and `-window-end=HH:MM`, that indicate a window per day during which Yale can perform key rotation operations.

### Motivation

When Yale rotates a key in broad-dsde-qa, it triggers a restart of all BEEs at once. This can be quite disruptive to developers and tests running in CI. Therefore we would like to configure key rotations for off-hours.

### Testing

This PR includes unit tests.

### CODE REVIEW

File: cmd/yale/main.go [hunk 1]
```diff
@@ -49,6 +63,69 @@ func parseArgs() *args {
 	local := flag.Bool("local", false, "use this flag when running locally (outside of cluster to use local kube config")
 	cacheNamespace := flag.String("cachenamespace", cache.DefaultCacheNamespace, "namespace where yale should cache service account keys")
 	ignoreUsageMetrics := flag.Bool("ignoreusagemetrics", false, "do not check if service account key is in use before disabling")
+	windowStart := flag.String("window-start", "", "use to restrict rotation to a particular time of day (HH:MM). eg. 05:00")
```
Comment (jyang-broad) [inline cmd/yale/main.go:66]: what's the timezone we're working w/ here ET or UT or system time (central time?)

File: internal/yale/yale.go [hunk 2]
```diff
@@ -149,6 +157,15 @@ func processYaleResource[Y apiv1b1.YaleCRD](yale *Yale, entry *cache.Entry, yale
 	if err = syncYaleResourceIfReady(yale.keysync, entry, yaleCRDs); err != nil {
 		return err
 	}
+
+	window := yale.options.RotateWindow
+	if window.Enabled {
+		if currentTime().Before(window.StartTime) || currentTime().After(window.EndTime) {
+			logs.Info.Printf("won't attempt key rotations for %s %s because we are outside the rotation window (%s - %s)", entry.Type, entry.Identifier, window.StartTime, window.EndTime)
```
Comment (jyang-broad) [inline internal/yale/yale.go:164]: not too familiar w/ yale here, but what's the retry logic/how do we guarantee we don't forever hit outside the window?

File: cmd/yale/main.go [hunk 3]
```diff
@@ -49,6 +63,69 @@ func parseArgs() *args {
...
+	hour, err := strconv.Atoi(hh)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse hour: %s", hh)
+	}
+	minute, err := strconv.Atoi(mm)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse minute: %s", mm)
+	}
+
+	t := time.Date(now.Year(), now.Month(), now.Day(), hour, minute, 0, 0, now.Location())
```
Comment (jyang-broad) [inline cmd/yale/main.go:129]: I think this might need additional logic to shove the date to tomorrow if WindowEnd < Window Start. i.e. 10pm <-> 2am

**Review by jyang-broad — COMMENTED** [review]
gave the draft a looksee. looks good! added some comments/questions re timezones / cross day windows.

File: internal/yale/yale.go (hunk 2 above)
Comment (choover-broad) [reply internal/yale/yale.go:164]: We will need to set sane windows that don't conflict with the Yale cronjob's scheduling. Eg. a 1-2 hour period once per day with a cronjob that runs every 1-5 minutes.

File: cmd/yale/main.go (hunk 1 above)
Comment (choover-broad) [reply cmd/yale/main.go:66]: Yale's local time, which will always be UTC on GKE K8s clusters.

**Review by mflinn-broad — APPROVED** [review]
stamped

**Pull request merged**